client/components/common/SelectElement/index.vue

```diff
@@ -2,88 +2,180 @@
-    allowedTypes: { type: Array, default: () => [] }
+    submitLabel: { type: String, default: 'Save' },
+    headerIcon: { type: String, default: 'mdi-toy-brick-plus-outline' },
+    useCurrentRepo: { type: Boolean, default: false }
+  },
+  data: () => {
+    return {
+      repository: null,
+      selectedActivity: null,
+      contentContainers: [],
+      selectedElements: [],
+      loadingContent: false,
+      activities: []
+    };
+  },
+  computed: {
+    ...mapGetters('repository', {
+      currentRepository: 'repository',
+      currentActivities: 'activities'
+    }),
+    allElementsSelected: vm => vm.selectedElements.length === vm.elements.length,
+    showBackButton: vm => vm.useCurrentRepo ? !!vm.selectedActivity : !!vm.repository,
+    containers() {
+      const { selectedActivity, activities } = this;
+      if (!selectedActivity || !activities.length) return [];
+      const rootTypes = getContainerTypes(selectedActivity.type);
+      let containers = activities.filter(({ type, parentId }) => {
+        return parentId === selectedActivity.id && rootTypes.includes(type);
+      });
+      containers = sortBy(containers, [
+        it => rootTypes.indexOf(it.type), 'position', 'createdAt'
+      ]);
+      return containers.reduce((acc, container) => {
+        const subcontainers = getContainers(activities, container);
+        acc.push(container, ...sortBy(subcontainers, 'position'));
+        return acc;
+      }, []);
+    },
+    elements() {
+      const elements = flatMap(this.contentContainers, 'elements');
+      if (!this.allowedTypes.length) return elements;
+      return elements.filter(it => this.allowedTypes.includes(it.type));
+    },
+    toggleButton() {
+      const { allElementsSelected, elements, multiple, selectedActivity } = this;
+      if (!multiple || !selectedActivity || !elements.length) return;
+      const { SELECT, DESELECT } = TOGGLE_BUTTON;
+      return allElementsSelected ? DESELECT : SELECT;
+    }
   },
-  data: () => ({
-    selectedActivity: null,
-    contentContainers: [],
-    selectedElements: [],
-    loadingContent: false
-  }),
-  computed: mapGetters('repository', ['repository', 'activities']),
   methods: {
     async showActivityElements(activity) {
       this.selectedActivity = activity;
-      const { activities } = this;
-      const containers = sortBy(getContainers(activities, activity), 'position');
+      const { containers } = this;
       const elements = await this.fetchElements(containers);
-      this.contentContainers = containers.map(container => ({
-        ...container,
-        elements: elements.filter(element => element.activityId === container.id)
-      }));
+      this.contentContainers = containers.map(container => {
+        const containerElements = elements.filter(it => it.activityId === container.id);
+        return { ...containers, elements: sortBy(containerElements, 'position') };
+      });
+    },
+    assignActivity(element) {
```

Comment:
This is used when the element is being toggled on or off which can theoretically happen frequently. Is there a reason this can not be called on all of the elements right after they are fetched and ordered by position?

Reply:
Thank you 😉